### DISCUSSION TITLE
Add a question and answer section in this repo

### DISCUSSION THREAD

**Discussion by jessgoldq4** [panva/node-oidc-provider#1270] [Ideas]
Why is there no where to have a discussion for clarification on the often-times unclear documentation, or the intended functionality of certain parts of the package? (i.e, why aren't access tokens stored in a redis store when using the example redis adapter; reading the code didn't make it clear either) 

I think it would be beneficial to have a place that's part of this repo to have these discussions or post issues

Thoughts?

**Comment by panva** [comment]
> I think it would be beneficial to have a place that's part of this repo to have these discussions or post issues

I do not have the time or energy to do Q&A support for this module. Once upon a time the issue / discussions tracker _was_ open for questions and free support. This only welcomed developers who, in my humble opinion, shouldn't attempt to do an identity provider of their own. As such, use the module as-is or don't, the goal of this module at this time is not to be a low barrier entry open source replacement for the plethora of feature-complete identity provider solutions that are available.

> why aren't access tokens stored in a redis store when using the example redis adapter;

Opaque Access Tokens _are_ stored in the adapter, where else would they be? Structured Access Tokens OTOH don't need to be stored.

**Reply by jessgoldq4** [reply]
I see. Thanks for the clarification. Totally understand that it takes a lot of effort to maintain communication here, especially if it's not the intended purpose of this repo.

> Structured Access Tokens OTOH don't need to be stored.

I was specifically asking about structured access tokens actually. Why don't they need to be stored? Just curious for my own knowledge

If it's because the access token's validity can be verified via jwks keys, then I assume when the session is invalidated, the tokens would be implicitly invalidated too?

---

_Feel free to cut the conversation short if you don't want to spend time answering_

**Reply by piixel13** [reply]
I can answer this with my limited knowledge of Oauth2.

Opaque token are just "random" strings, so whenever a service provider uses them to claim access to a resource, the IDP needs to map this "random" string to some information (the user that is authenticated, the SP that this token has been granted to, the audiance, the scope, expiration ...etc), so we need to store the opaque token the IDP side to be able to retreive the information associated to it

OTOH, Structured JWT token contain all of this information in their payload, and this payload is signed by the IDP itself (with it's own private key, so only decryptable by it's associated public key), that way when an SP uses the JWT token to claim access to some resource, the IDP just needs to decode the token, verify it's signature (and some other params), and then if the verification passes, the IDP can just use the information contained in the JWT payload, since it was signed, the IDP can be sure that it is the issuer of this token, that the information in the payload hasn't been tampered with (because the IDP signs the hash of the payload), that the JWT isn't expired ...etc

This is why it isn't required by the Oauth2 standard to persist the JWT access token, but this is only for Oauth2, there may be some use cases where we need to store JWT access token (revocation, caching ...etc) but this falls outside of the scope of oauth2 and oidc, that's why this lib dosn't provide it

I hope that my answer is clear, and doesn't contain ( too many :) ) uncorrect information, @panva i invite you to correct or add information if you have the time and the desire to ^^

**Reply by jessgoldq4** [reply]
Hi @Tiieb, thanks for the reply. That aligns with my understanding. Still curious on how it's handled when the session is expired

**Reply by panva** [reply]
Just one correcting point: this particular IdP module does not accept its own JWT structured tokens for any purpose. A structured JWT AT is always issued for a particular resource server which is not the IdP itself. So the JWT AT is either signed by the IdP and verifiable by the resource server it's accepted at, or it's encrypted for the resource server. Either way, there's no use for the AT at this IdP.

> Hi @Tiieb, thanks for the reply. That aligns with my understanding. Still curious on how it's handled when the session is expired

It isn't. JWT structured tokens are self contained. Their status doesn't change throughout their lifetime.

**Reply by jessgoldq4** [reply]
Huh, curious. So even if the session is invalidated, the JWT is still alive... curious. I wonder if there's common strategies for revoking the token in that case. I can look into that on my own. Thanks @panva, appreciate it

**Reply by panva** [reply]
The point of a JWT AT is so that the resource server does not need to query the IdP for the status or contents of its authorization. Having the ability to do so would completely void the point of the JWT AT, at that point, use an opaque AT.

**Reply by JonasDoe** [reply]
Regarding the original topic: while I personally wished there was some Q&A section, I understand why it's not. That said, since this oidc-provider is most likely an enterprise solution, it's fair that you ask your company to [support](https://github.com/panva/node-oidc-provider?tab=readme-ov-file#support) it. With that, you'll receive a more professional support.
I think we all know github projects of committed individuals which went down at some point b/c the maintainer suffered a breakdown from working 24/7 for it, for free. If this sponsoring is the most viable approach to prevent that - so be it :man_shrugging:.
